kousa/lib/business-logic/room_chat_logic.ex

```diff
@@ -1,33 +1,97 @@
-    case Data.User.get_current_room_id(user_id) do
-      nil ->
-        nil
+  @message_character_limit 512
 
-      current_room_id ->
-        with {avatar_url, display_name} <-
-               Gen.UserSession.send_call!(user_id, {:get_info_for_msg}) do
-          RegUtils.lookup_and_cast(
-            Gen.RoomChat,
-            current_room_id,
-            {:new_msg, user_id,
-             %{
-               id: Ecto.UUID.generate(),
-               avatarUrl: avatar_url,
-               displayName: display_name,
-               userId: user_id,
-               tokens: tokens
-             }}
-          )
-        end
+  @spec send_msg(String.t(), list(map)) :: any
+  def send_msg(user_id, tokens) do
+    tokens = validate_tokens(tokens)
+    
+    if length(tokens) > 0 do
+      case Data.User.get_current_room_id(user_id) do
+        nil ->
+          nil
+
+        current_room_id ->
+          with {avatar_url, display_name} <-
+                Gen.UserSession.send_call!(user_id, {:get_info_for_msg}) do
+            RegUtils.lookup_and_cast(
+              Gen.RoomChat,
+              current_room_id,
+              {:new_msg, user_id,
+              %{
+                id: Ecto.UUID.generate(),
+                avatarUrl: avatar_url,
+                displayName: display_name,
+                userId: user_id,
+                tokens: tokens
+              }}
+            )
+          end
+      end
+    end
+  end
+
+  defp validate_tokens(tokens) when is_list(tokens) do
+    if Enum.reduce_while(tokens, 0, &count_message_characters/2) <= @message_character_limit do
+      tokens 
+      |> Enum.reduce([], &validate_tokens/2)
+      |> Enum.reverse()
+    else
+      []
     end
   end
 
-  def send_msg(_user_id, _tokens) do
+  defp validate_tokens(_), do: []
+
+  defp validate_tokens(token, acc) do
+    case validate_token(token) do
+      {:ok, token} -> [token | acc]
+      _ -> acc
+    end
+  end
+
+  defp count_message_characters(%{"v" => v}, acc) do
+    if acc <= @message_character_limit, do: {:cont, String.length(v) + acc}, else: {:halt, acc}
+  end
+
+  @doc """
+    validate_token/1 validates each token by type and return {:ok, token} if the token is valid
+    otherwise return :invalid
+  """
+  defp validate_token(token = %{"t" => "text", "v" => _}), do: {:ok, token}
+  defp validate_token(token = %{"t" => "mention", "v" => mention}) do 
+    mention
+    |> valid_mention?()
+    |> case do
+      true -> {:ok, token}
+      _ -> :invalid
+    end
+  end
+
+  defp validate_token(token = %{"t" => "link", "v" => link}) do
+    link
+    |> URI.parse() 
+    |> valid_url?()
+    |> case do
+      true -> {:ok, token}
+      _ -> :invalid
+    end
+  end
+
+  defp validate_token(_), do: :invalid
+
+  defp valid_url?(%URI{host: host, scheme: scheme}) when is_binary(host) and is_binary(scheme), do: true
+  defp valid_url?(_), do: false
+
+  defp valid_mention?(mention) do
+    mention
+    |> Data.User.get_by_username()
```

Comment:
I want to avoid a db lookup on message sends, I'm ok if someone does an @ for a non-existent user, so you can remove this

Reply:
it won't work for a non-existing user anyway.

Reply:
@benawad @nadirabbas yea that sounds reasonable according to business. I will remove it.